### PR TITLE
Replace asserts with errors, resolves #242

### LINKED ISSUE #242: Assert Statements are for Debugging
Currently MHKiT uses assert staements in its production code (primarilty for type checking functional inputs). As stated in the docs (https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement) assert stements are for debugging. 

We should raise an error if the user does not pass the correct information. 

So going forward we would convert:
```python
assert isinstance(station_number, (str, type(None))), (f'station_number must be  of type string')
```
to this

```python
if not isinstance(station_number, (str, type(None))):
    raise ValueError(f'station_number must be of type string. Got: {station_number}')
```

### PR DESCRIPTION
**Pull request by akeeste**
This PR replaces all non-testing asserts with code that will raise errors. For the vast majority of asserts, the error messages are unchanged or have additions to say the parameter type or value that was given and is causing an issue.

Some try-except statements appeared like they should be errors, but just defined unused strings. I updated these to throw error messages.

Modules completed:
- [x] loads
- [x] power
- [x] utils
- [x] tidal 
- [x] dolfyn
- river/io/d3d
- [x] river
- [x] wave/io
- [x] wave

I left ``river/io/d3d.py`` alone for now because #271 edits it and I wanted to avoid conflicts if possible.

Other items
- [x] @akeeste confirm that the new exceptions being thrown are caught by tests that evaluate incorrect input
- [x] I also changed a parameter in ``plot_directional_spectrum``. It was using a parameter ``min`` that overwrote the built in Python function. i changed the parameter to ``color_level_min`` to prevent potential issues.

### CODE REVIEW

File: mhkit/wave/io/hindcast/hindcast.py [hunk 1]
```diff
@@ -330,13 +338,18 @@ def request_wpto_directional_spectrum(
...
-    assert isinstance(tree, (str, type(None))), 'tree must be a sring'
-    assert isinstance(unscale, bool), 'unscale must be bool type'
-    assert isinstance(str_decode, bool), 'str_decode must be bool type'
-    assert isinstance(hsds, bool), 'hsds must be bool type'
+    if not isinstance(lat_lon, (list, tuple)):
+        raise TypeError(f'lat_lon must be of type list or tuple. Got: {type(lat_lon)}')
+    if not isinstance(year, str):
+        raise TypeError(f'years must be a string. Got: {type(year)}')
+    if not isinstance(tree, (str, type(None))):
+        raise TypeError(f'tree must be a sring. Got: {type(tree)}')
```
Comment (ssolson) [inline mhkit/wave/io/hindcast/hindcast.py:346]: "sring"=>string

**Review by ssolson — COMMENTED** [review]
Adam thanks for knocking this out quickly. Couple of comments to address before merging.

File: mhkit/loads/extreme.py [hunk 2]
```diff
@@ -707,8 +734,9 @@ def mler_simulation(parameters=None):
         Simulation parameters including spatial and time calculated
         arrays.
     """
-    if not parameters == None:
-        assert isinstance(parameters, dict), 'parameters must be of type dict'
+    if not parameters == None and not isinstance(parameters, dict):
```
Comment (akeeste) [inline mhkit/loads/extreme.py:737]: @akeeste fix this optional argument

File: mhkit/wave/contours.py [hunk 3]
```diff
@@ -1504,12 +1543,14 @@ def _bivariate_KDE(x1, x2, bw, fit, nb_steps, Ndata_bivariate_KDE, kwargs):
         max_x1 = x1.max()*2
     if isinstance(max_x2, type(None)):
         max_x2 = x2.max()*2
```
Comment (akeeste) [inline mhkit/wave/contours.py:1545]: @ssolson I'm not familiar with this method, but I noticed this line doesn't seem to match the description. Is it accurate to set ``max_x1 = 2*x1.max()``?

File: mhkit/wave/resource.py [hunk 4]
```diff
@@ -215,41 +228,48 @@ def surface_elevation(S, time_index, seed=None, frequency_bins=None, phases=None
...
-            'phases must be of type None, np.ndarray, or pd,DataFrame')
-    assert isinstance(method, str)
+    if not isinstance(S, pd.DataFrame):
+        raise TypeError(f'S must be of type pd.DataFrame. Got: {type(S)}')
+    if not isinstance(time_index, np.ndarray):
+        raise TypeError(f'time_index must be of type np.ndarray. Got: {type(time_index)}')
+    if not isinstance(seed, (type(None),int)):
+        raise TypeError(f'seed must be of type int. Got: {type(seed)}')
+    if not isinstance(frequency_bins, (type(None), np.ndarray, pd.DataFrame)):
+        raise TypeError('frequency_bins must be of type None, np.ndarray, or pd, DataFrame. Got: {type(frequency_bins)}')
```
[the remark this answers is not in the document]
Comment (akeeste) [reply mhkit/wave/resource.py]: I prefer the "If specified, ..." option too. I implemented that language for all the parameters that default to None

File: mhkit/wave/contours.py (hunk 3 above)
Comment (ssolson) [reply mhkit/wave/contours.py:1543]: This comes directly from WDRT:

https://github.com/WEC-Sim/WDRT/blob/71b2b871d27950bf14a944f432f8ee1221ee05e3/WDRT/ESSC.py#L2552-L2585
image

**Comment by ssolson** [comment]
@akeeste just checking if this is ready for another review

**Comment by akeeste** [comment]
> @akeeste just checking if this is ready for another review

@ssolson Yes. I implemented all of the suggested changes. Let me know if there's any other changes you'd like to see. 

I'm debating if the update to the ``min`` parameter of ``plot_directional_spectrum`` is necessary. How much of an issue is it to overwrite the python function ``min``? Seems safer to not overwrite it, but the function call won't be backwards compatible then.

**Comment by ssolson** [comment]
> I'm debating if the update to the `min` parameter of `plot_directional_spectrum` is necessary. How much of an issue is it to overwrite the python function `min`? Seems safer to not overwrite it, but the function call won't be backwards compatible then.

Its a very easy fix for the user to change a kwarg from `min` to `color_bar_min` and is the correct way to handle.

File: mhkit/wave/resource.py [hunk 5]
```diff
@@ -36,15 +36,22 @@ def elevation_spectrum(eta, sample_rate, nnft, window='hann',
 
     # TODO: Add confidence intervals, equal energy frequency spacing, and NDBC
     #       frequency spacing
-    # TODO: may need an assert for the length of nnft- signal.welch breaks when nfft is too short
+    # TODO: may need to raise an error for the length of nnft- signal.welch breaks when nfft is too short
     # TODO: check for uniform sampling
```
Comment (ssolson) [inline mhkit/wave/resource.py:40]: Unrelated to this PR and mostly a note to self but we shouldn't have TODOs in production code. These should be captured in issues and removed.

**Review by ssolson — APPROVED** [review]
Adam I found one file where a copy pasta "be" was excluded from multiple exceptions. And I found one place where "If specified" should be added.

I added a note to self about the TODOs but lets not address that here.

After you fix these minor issues above let's merge.

File: mhkit/utils/upcrossing.py [hunk 6]
```diff
@@ -108,11 +107,11 @@ def peaks(t, data, inds=None):
...
 
     """
+    # Check data types
     if not isinstance(t, np.ndarray):
-        TypeError('t must be of type np.ndarray')
-        
+        raise TypeError(f't must be of type np.ndarray. Got: {type(t)}')
     if not isinstance(data, np.ndarray):
-        TypeError('data must be of type np.ndarray')
+        raise TypeError(f'data must of type np.ndarray. Got: {type(data)}')
```
[the remark this answers is not in the document]
Comment (akeeste) [reply mhkit/utils/upcrossing.py:114]:
```suggestion
        raise TypeError(f'data must be of type np.ndarray. Got: {type(data)}')

File: mhkit/utils/upcrossing.py [hunk 7]
```diff
@@ -139,8 +138,11 @@ def troughs(t, data, inds=None):
         Trough values of the time-series
 
     """
-    assert isinstance(t, np.ndarray), 't must be of type np.ndarray'
-    assert isinstance(data, np.ndarray), 'data must be of type np.ndarray'
+    # Check data types
+    if not isinstance(t, np.ndarray):
+        raise TypeError(f't must be of type np.ndarray. Got: {type(t)}')
+    if not isinstance(data, np.ndarray):
+        raise TypeError(f'data must of type np.ndarray. Got: {type(data)}')
```
[the remark this answers is not in the document]
Comment (akeeste) [reply mhkit/utils/upcrossing.py:145]:
```suggestion
        raise TypeError(f'data must be of type np.ndarray. Got: {type(data)}')

File: mhkit/utils/upcrossing.py [hunk 8]
```diff
@@ -169,8 +171,11 @@ def heights(t, data, inds=None):
     heights: np.array
         Height values of the time-series
     """
-    assert isinstance(t, np.ndarray), 't must be of type np.ndarray'
-    assert isinstance(data, np.ndarray), 'data must be of type np.ndarray'
+    # Check data types
+    if not isinstance(t, np.ndarray):
+        raise TypeError(f't must be of type np.ndarray. Got: {type(t)}')
+    if not isinstance(data, np.ndarray):
+        raise TypeError(f'data must of type np.ndarray. Got: {type(data)}')
```
[the remark this answers is not in the document]
Comment (akeeste) [reply mhkit/utils/upcrossing.py:178]:
```suggestion
        raise TypeError(f'data must be of type np.ndarray. Got: {type(data)}')

File: mhkit/utils/upcrossing.py [hunk 9]
```diff
@@ -199,8 +204,11 @@ def periods(t, data, inds=None):
     periods: np.array
         Period values of the time-series
     """
-    assert isinstance(t, np.ndarray), 't must be of type np.ndarray'
-    assert isinstance(data, np.ndarray), 'data must be of type np.ndarray'
+    # Check data types
+    if not isinstance(t, np.ndarray):
+        raise TypeError(f't must be of type np.ndarray. Got: {type(t)}')
+    if not isinstance(data, np.ndarray):
+        raise TypeError(f'data must of type np.ndarray. Got: {type(data)}')
```
[the remark this answers is not in the document]
Comment (akeeste) [reply mhkit/utils/upcrossing.py:211]:
```suggestion
        raise TypeError(f'data must be of type np.ndarray. Got: {type(data)}')

File: mhkit/utils/upcrossing.py [hunk 10]
```diff
@@ -230,9 +238,12 @@ def custom(t, data, func, inds=None):
...
     """
-
-    assert isinstance(t, np.ndarray), 't must be of type np.ndarray'
-    assert isinstance(data, np.ndarray), 'data must of type np.ndarray'
-    assert callable(func), 'f must be callable'
+    # Check data types
+    if not isinstance(t, np.ndarray):
+        raise TypeError(f't must be of type np.ndarray. Got: {type(t)}')
+    if not isinstance(data, np.ndarray):
+        raise TypeError(f'data must of type np.ndarray. Got: {type(data)}')
```
[the remark this answers is not in the document]
Comment (akeeste) [reply mhkit/utils/upcrossing.py:245]:
```suggestion
        raise TypeError(f'data must be of type np.ndarray. Got: {type(data)}')

File: mhkit/wave/resource.py [hunk 11]
```diff
@@ -133,11 +143,14 @@ def jonswap_spectrum(f, Tp, Hs, gamma=None):
...
-    assert isinstance(gamma, (int,float, type(None))), \
-        'gamma must be of type int or float'
+    if not isinstance(f, np.ndarray):
+        raise TypeError(f'f must be of type np.ndarray. Got: {type(f)}')
+    if not isinstance(Tp, (int,float)):
+        raise TypeError(f'Tp must be of type int or float. Got: {type(Tp)}')
+    if not isinstance(Hs, (int,float)):
+        raise TypeError(f'Hs must be of type int or float. Got: {type(Hs)}')
+    if not isinstance(gamma, (int,float, type(None))):
+        raise TypeError(f'gamma must be of type int or float. Got: {type(gamma)}')
```
Comment (akeeste) [inline mhkit/wave/resource.py:153]:
```suggestion
        raise TypeError(f'If specified, gamma must be of type int or float. Got: {type(gamma)}')

**Comment by akeeste** [comment]
Thanks @ssolson. I resolved those last few issues and will merge

**Pull request merged**